enumerables.rb

```diff
@@ -0,0 +1,137 @@
+# rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity, Metrics/ModuleLength
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    while i < size
+      if is_a? Array
+        yield self[i]
+      elsif is_a? Hash
+        yield keys[i], self[keys[i]]
+      elsif is_a? Range
+        yield to_a[i]
+      end
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    while i < size
+      yield to_a[i], i
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    result = []
+    to_a.my_each { |e| result << e if yield e }
+    result
+  end
+
+  def my_all?(arg = nil)
+    result = true
+    if block_given?
+      my_each { |e| result = false unless yield e }
+    elsif arg.is_a?(Class) && arg.class != Regexp
+      my_each { |e| result = false unless e.is_a?(arg) }
+    elsif arg.is_a?(Regexp)
+      my_each { |e| result = false unless e.match(arg) }
+    elsif arg.nil?
+      block = proc { |e| e.nil? || e == false }
+      my_each { |_e| result = false if block.call }
+    else
+      block = proc { |e| e == arg }
+      my_each { |_e| result = false unless block.call }
+    end
+    result
+  end
```

Comment:
#my_all? when no block or argument is given returns true when none of the collection members are false or nil
#my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern